{% extends "layout.html" %}

<!DOCTYPE html>
<html>
<head>
  <!-- Your head content here (e.g., title, meta tags) -->
  <style>
    body {
      font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
      line-height: 1.6;
      color: #252525;
      padding: 20px;
      background-color: #f8f8f8;
    }
  
    .content {
      background-color: #fff;
      padding: 20px;
      margin: 20px auto;
      max-width: 800px;
      box-shadow: 0 2px 4px rgba(0,0,0,0.1);
      font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
    }
  
    h2 {
      color: #252525;
      border-bottom: 1px solid #eee;
      padding-bottom: 10px;
      font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
    }
  
    p {
      margin-bottom: 16px; /* adds space between paragraphs */
      text-align: justify;
      line-height:2; /* optional: justifies the text of paragraphs */
    }
  
    /* You can add more styles as needed */
  </style>
  
</head>


<body>

{% block content %}
<div class="content">
  <h2 style="font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif; text-align: center; font-size: 50px;">ADMIXTURE ANALYSIS</h2>
  <br> 
  <p>
    Admixture analysis is used to infer the ancestry of individuals by determining the proportion of an individual's ancestry that comes from each of several potential source populations. This can provide insights into the historical migrations and mixtures of populations.
  </p>
  <br>
  <p>
    This analysis can be crucial for understanding the genetic structure of populations and for studying the genetic basis of diseases. By identifying the ancestral proportions in an individual's genome, researchers can better account for genetic variation in association studies.
  </p>
  <br>
  <p>
    Admixture analysis is used to infer the ancestry of individuals by determining the proportion of an individual's ancestry that comes from each of several potential source populations. This can provide insights into the historical migrations and mixtures of populations. This analysis can be crucial for understanding the genetic structure of populations and for studying the genetic basis of diseases. By identifying the ancestral proportions in an individual's genome, researchers can better account for genetic variation in association studies.
  </p>
  <br>
  <p>
    ADMIXTURE is caused by interbreeding between isolated populations and this results in a mixture of alleles from different ancestral populations. Analysis of this allows classification of individuals with unknown ancestry into discrete populations [1]. ‘Ancestral’ or ‘parental’ refers to the previously isolated populations, whilst ‘admixed’ refers to the new population [2].
  </p>
  <br>
  <p>
    Several software tools, such as STRUCTURE, and ADMIXTURE facilitate these analyses by employing sophisticated algorithms designed to handle large genomic datasets. The outcomes of admixture analysis are pivotal for understanding genetic diversity, mapping genetic diseases, and providing ancestry information in personal genetic testing. Bayesian principle of likelihood comparison is used for STRUCTURE and prior information can be used to support further shaping of analysis. A similar method is used for ADMIXTURE but the population parameter calculation is a lot quicker and therefore allows for use of much larger marker sets. However, linkage disequilibrium is not taken into account for ADMIXTURE [3].
  </p>
  <br>
  <p>For more details about results and how we obtained them, please refer to our documentation.</p>
  <br>
  <p style="font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif; font-size: 30px">
    REFERENCES
  </p>
  <br>
  <p>

    [1] Skotte L, Korneliussen TS, Albrechtsen A. Estimating individual admixture proportions from next generation sequencing data. Genetics. 2013 Nov 1;195(3):693–702. doi:<a href="https://doi.org/10.1534/genetics.113.154138" target="_blank">https://doi.org/10.1534/genetics.113.154138</a> 
  </p>
  <br>
  <p>
    [2] Shriner D. Overview of admixture mapping. Current Protocols in Human Genetics. 2013 Jan;76(1). doi:<a href="https://doi.org/10.1002/0471142905.hg0123s76" target="_blank"> https://doi.org/10.1002/0471142905.hg0123s76</a>
  </p>
  <br>
  <p>
    [3] Porras-Hurtado L, Ruiz Y, Santos C, Phillips C, Carracedo Á, Lareu MV. An overview of structure: Applications, parameter settings, and supporting software. Frontiers in Genetics. 2013 May 29;4. doi: <a href="https://doi.org/10.3389%2Ffgene.2013.00098" target="_blank">https://doi.org/10.3389%2Ffgene.2013.00098</a>

  </p>
  <!-- You can add more content here based on your project's needs -->
</div>


</body>






<link rel="stylesheet" href="{{ url_for('static', filename='css/footer.css') }}">
<div class="site-footer">
    <footer class="bottom-banner"> 
        <div class="footer-content">
            <div class="social-media">
                <a href="https://github.com/Djinho/Population-Genetics-Web-Tool" target="_blank" rel="noopener noreferrer">
                    <img src="{{ url_for('static', filename='images/github-mark-white.png') }}">
                </a>
            </div>
            <div class="copyright">
                <p>&copy; 2024 GenomePop Analytics. All rights reserved.</p>
            </div>

        </div>
    </footer>
{% endblock %}
</html>