{% extends "layout.html" %}

<!DOCTYPE html>
<html>
<head>
  <!-- Your head content here (e.g., title, meta tags) -->
  
</head>

{% block content %}
<div class="content">
    <h2 clas="tutorial-head">SNP ANALYSIS</h2>
    <br>
    <p>Humans mostly have identical genomes therefore only a small proportion of genetic differences have caused the immense variation seen in individuals. All of these changes are due to genetic differences between individuals therefore identifying and observing the single base changes, also known as single nucleotide polymorphisms (SNPs), is ideal [1].</p>
    <br> 
    <p>The fixation index (FST)  is used to measure how these populations differ based on their genetics. For our FST values, we used the general formula made by Sewell Wright [2]. 
        This formula is described as: FST = s^2/¯p(1-¯p), where s^2 is the sample variance of allele frequencies and ¯p  is the mean [3]. 
    </p>
    <br>
    <p>
        Our method of calculating FST, however, involved the use of weighted averages and this formula is: FST= (H_T-H_S)/H_T . H_T  represents the total or expected heterozygosity in the entire population and it is calculated as 2¯p (1-¯p) where ¯p represents the weighted average allele frequency across all SNPs for the populations being compared. H_S represents the weighted average of within-population heterozygosity and is derived from the weighted average mean of 2p_pop (1-p_pop) across all SNPs within each population. The p_pop represents the allele frequency in each population and the weights are the sample sizes of the populations.
    </p>
    <br>
    <p>
        Prior to implementing this formula, we aggregated the SNP data across populations and this reduced the complexity, whilst increasing accuracy through the use of sample sizes as ‘weights’.  
    </p>
    <br>
    <p>For more details about results and how we obtained them, please refer to our documentation.</p>
    <br>
    <p class="tutorial-h3">REFERENCES</p>
    <br>
    <p>[1] Norrgard K, Schultz J. Using SNP Data to Examine Human Phenotypic Differences [Internet]. Nature Publishing Group; 2008 [cited 2024 Feb 27]. Available from: <a href="https://www.nature.com/scitable/topicpage/using-snp-data-to-examine-human-phenotypic-706/ " target="_blank">https://www.nature.com/scitable/topicpage/using-snp-data-to-examine-human-phenotypic-706/ </a> </p>
    <br>
    <p>[2]  Wright S. The interpretation of population structure by F-statistics with special regard to systems of mating. Evolution. 1965 Sept;19(3):395. doi: <a href=" https://doi.org/10.2307/2406450" target="_blank"> https://doi.org/10.2307/2406450</a> </p>
    <br>
    <p>[3] Weir BS. Estimating f-statistics: A historical view. Philosophy of Science. 2012 Dec;79(5):637–43. doi: <a href="https://doi.org/10.1086%2F667904" target="_blank">https://doi.org/10.1086%2F667904</a>  </p>




<link rel="stylesheet" href="{{ url_for('static', filename='css/footer.css') }}">
<div class="site-footer">
    <footer class="bottom-banner"> 
        <div class="footer-content">
            <div class="social-media">
                <a href="https://github.com/Djinho/Population-Genetics-Web-Tool" target="_blank" rel="noopener noreferrer">
                    <img src="{{ url_for('static', filename='images/github-mark-white.png') }}">
                </a>
            </div>
            <div class="copyright">
                <p>&copy; 2024 GenomePop Analytics. All rights reserved.</p>
            </div>

        </div>
    </footer>
</div>


{% endblock %}
