{% extends "layout.html" %}

<!DOCTYPE html>
<html>
<head>
  <!-- Your head content here (e.g., title, meta tags) -->
  <style>
    body {
      font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
      line-height: 1.6;
      color: #252525;
      padding: 20px;
      background-color: #f8f8f8;
    }
  
    .content {
      background-color: #fff;
      padding: 20px;
      margin: 20px auto;
      max-width: 800px;
      box-shadow: 0 2px 4px rgba(0,0,0,0.1);
      font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
    }
  
    h2 {
      color: #252525;
      border-bottom: 1px solid #eee;
      padding-bottom: 10px;
      font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
    }
  
    p {
      margin-bottom: 16px; /* adds space between paragraphs */
      text-align: justify;
      line-height:2; /* optional: justifies the text of paragraphs */
    }
  
    /* You can add more styles as needed */
  </style>
  
</head>

{% block content %}
<div class="content">
    <h2 style="font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif; font-size: 50px; text-align: center;">SNP ANALYSIS</h2>
    <br>
    <p>Humans mostly have identical genomes therefore only a small proportion of genetic differences have caused the immense variation seen in individuals. All of these changes are due to genetic differences between individuals therefore identifying and observing the single base changes, also known as single nucleotide polymorphisms (SNPs), is ideal [1].</p>
    <br> 
    <p>The fixation index (FST)  is used to measure how these populations differ based on their genetics. For our FST values, we used the general formula made by Sewell Wright [2]. 
        This formula is described as: FST = s^2/¯p(1-¯p), where s^2 is the sample variance of allele frequencies and ¯p  is the mean [3]. We calculated weighted averages using our sample size without incorporating any covariates, focusing solely on allele frequencies. We then computed the average allele frequency across all single nucleotide polymorphisms (SNPs) and applied a single FST calculation for the entire dataset.
    </p>
    <br>
    <p>For more details about results and how we obtained them, please refer to our documentation.</p>
    <br>
    <p style="font-family: 'Bebas Neue', Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif; font-size: 30px; text-align: center;">REFERENCES</p>
    <br>
    <p>[1] Norrgard K, Schultz J. Using SNP Data to Examine Human Phenotypic Differences [Internet]. Nature Publishing Group; 2008 [cited 2024 Feb 27]. Available from: <a href="https://www.nature.com/scitable/topicpage/using-snp-data-to-examine-human-phenotypic-706/ " target="_blank">https://www.nature.com/scitable/topicpage/using-snp-data-to-examine-human-phenotypic-706/ </a> </p>
    <br>
    <p>[2]  Wright S. The interpretation of population structure by F-statistics with special regard to systems of mating. Evolution. 1965 Sept;19(3):395. doi: <a href=" https://doi.org/10.2307/2406450" target="_blank"> https://doi.org/10.2307/2406450</a> </p>
    <br>
    <p>[3] Weir BS. Estimating f-statistics: A historical view. Philosophy of Science. 2012 Dec;79(5):637–43. doi: <a href="https://doi.org/10.1086%2F667904" target="_blank">https://doi.org/10.1086%2F667904</a>  </p>




<link rel="stylesheet" href="{{ url_for('static', filename='css/footer.css') }}">
<div class="site-footer">
    <footer class="bottom-banner"> 
        <div class="footer-content">
            <div class="social-media">
                <a href="https://github.com/Djinho/Population-Genetics-Web-Tool" target="_blank" rel="noopener noreferrer">
                    <img src="{{ url_for('static', filename='images/github-mark-white.png') }}">
                </a>
            </div>
            <div class="copyright">
                <p>&copy; 2024 GenomePop Analytics. All rights reserved.</p>
            </div>

        </div>
    </footer>
</div>


{% endblock %}
